% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/ERVmap.R, R/TEtranscripts.R, R/Telescope.R,
%   R/qtex.R
\name{qtex,ERVmapParam-method}
\alias{qtex,ERVmapParam-method}
\alias{qtex}
\alias{qtex,TEtranscriptsParam-method}
\alias{qtex,TelescopeParam-method}
\alias{qtex,AtenaParam-method}
\title{Quantify transposable element expression}
\usage{
\S4method{qtex}{ERVmapParam}(
  x,
  phenodata = NULL,
  mode = ovUnion,
  yieldSize = 1000000L,
  verbose = 1,
  BPPARAM = SerialParam(progressbar = ifelse(verbose == 1, TRUE, FALSE)),
  on.disk = FALSE
)

\S4method{qtex}{TEtranscriptsParam}(
  x,
  phenodata = NULL,
  mode = ovUnion,
  yieldSize = 1000000L,
  BPPARAM = SerialParam(progressbar = TRUE),
  on.disk = FALSE
)

\S4method{qtex}{TelescopeParam}(
  x,
  phenodata = NULL,
  mode = ovUnion,
  yieldSize = 1000000L,
  BPPARAM = SerialParam(progressbar = TRUE)
)
}
\arguments{
\item{x}{An \code{AtenaParam} object of one of the following
subclasses:
\itemize{
  \item A \code{ERVmapParam} object built using the constructor
        function \code{\link{ERVmapParam}()}. This object will
        trigger \code{qtex()} to use the algorithm by
        Tokuyama et al. (2018).
  \item A \code{TelescopeParam} object built using the constructor
        function \code{\link{TelescopeParam}()}. This object will
        trigger \code{qtex()} to use the algorithm by
        Bendall et al. (2019).
}}

\item{phenodata}{A \code{data.frame} or \code{DataFrame} object storing
phenotypic data to include in the resulting
\code{SummarizedExperiment} object. If \code{phenodata} is set,
its row names will become the column names of the resulting
\linkS4class{SummarizedExperiment} object.}

\item{mode}{One of the pre-defined overlapping methods such as
\code{ovUnion()}, \code{ovIntersectionStrict} or a user-supplied
overlapping function. For a user-supplied overlapping function, the
input parameters must match those of the pre-defined methods and
the function must return a \code{\link[S4Vectors]{Hits}} object with
subject hits matching the annotated features. This parameter is
analogous to the \code{mode} parameter of the
\code{\link[GenomicAlignments]{summarizeOverlaps}()} function from
the \code{GenomicAlignments} package.}

\item{yieldSize}{Field inherited from \code{\link[Rsamtools]{BamFile}}.
The method for signature \code{\link{ERVmapParam}()} reads the BAM
file by chunks. \code{yieldSize} represents the number of records
(chunk size) to yield each time the file is read.}

\item{verbose}{(Default 1). When \code{verbose} > 1, detailed information on
the quantification steps is provided. Warnings are always present
regardless of the value of \code{verbose}.}

\item{BPPARAM}{An object of a \linkS4class{BiocParallelParam} subclass
to configure the parallel execution of the code. By default,
a \linkS4class{SerialParam} object is used, which does not use
any parallelization, with the flag \code{progress=TRUE} to show
progress through the calculations.}
}
\value{
A \linkS4class{SummarizedExperiment} object.
}
\description{
The \code{qtex()} method quantifies transposable element expression.
}
\details{
Giving some \code{AtenaParam} object sub-class as input, the
\code{qtex()} method quantifies the expression of transposable
elements (TEs). The particular algorithm to perform the
quantification will be selected depending on the specific
sub-class of input \code{AtenaParam} object, see argument
\code{x} above.
}
\examples{
bamfiles <- list.files(system.file("extdata", package="atena"),
                       pattern="*.bam", full.names=TRUE)
TE_annot <- readRDS(file = system.file("extdata", "Top28TEs.rds",
                    package="atena"))
gene_annot <- readRDS(file = system.file("extdata", "Top50genes.rds",
                                         package="atena"))
tspar <- TelescopeParam(bfl=bamfiles, teFeatures=TE_annot,
                        geneFeatures = gene_annot,
                        singleEnd = TRUE, ignoreStrand=TRUE)
tsSE <- qtex(tspar)
}
\references{
Tokuyama M et al. ERVmap analysis reveals genome-wide transcription of human
endogenous retroviruses. PNAS, 115(50):12565-12572, 2018.
\url{https://doi.org/10.1073/pnas.1814589115}

Bendall ML et al. Telescope: characterization of the retrotranscriptome by
accurate estimation of transposable element expression.
PLOS Computational Biology, 15:e1006453, 2019.
\url{https://doi.org/10.1371/journal.pcbi.1006453}
}
\seealso{
\code{\link{ERVmapParam}}
\code{\link{TelescopeParam}}
}
